### PR TITLE
Always fetch authentication tokens in Kerberos granted UGI

### PR DESCRIPTION
**Pull request by jerryshao**
CC @mridulm @bikassaha

### CODE REVIEW

**Comment by mridulm** [comment]
I am not sure if we can assume we will always have principal/keytab.
@bikassaha any comments ?

**Comment by jerryshao** [comment]
If not, then I don't know there's a way to address it.

**Comment by mridulm** [comment]
@jerryshao I am trying to find a workaround ... if that does not work, we will go with this as fallback

**Comment by jerryshao** [comment]
@mridulm I think the keypoint is to get a UGI with only tgt granted, so HBase client will honor this tgt to get new tokens.

**Comment by jerryshao** [comment]
Let me merge this PR first, @weiqingy you need to change the code of #120  based on this.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseCredentialsManager.scala [hunk 1]
```diff
@@ -44,6 +47,11 @@ final class HBaseCredentialsManager private() extends Logging {
     Executors.newSingleThreadScheduledExecutor(
       ThreadUtils.namedThreadFactory("HBase Tokens Refresh Thread"))
 
+  require(sparkConf.contains("spark.yarn.principal"))
+  require(sparkConf.contains("spark.yarn.keytab"))
+  private val principal = sparkConf.get("spark.yarn.principal")
+  private val keytab = sparkConf.get("spark.yarn.keytab")
+
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseCredentialsManager.scala]: Based on comments by @bikassaha , we are supposed to have shc specific principal/keytab and falling back to spark config if unavailable.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseCredentialsManager.scala [hunk 2]
```diff
@@ -73,7 +81,7 @@ final class HBaseCredentialsManager private() extends Logging {
       }
       logInfo(s"Obtain new token for cluster $identifier")
 
-      credentials.addToken(tokenInfo.token.getService, tokenInfo.token)
+      credentials.addToken(new Text(identifier), tokenInfo.token)
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseCredentialsManager.scala]: This is incorrect, as discussed yesterday, this must be `token.getService`

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseCredentialsManager.scala [hunk 3]
```diff
@@ -130,12 +138,17 @@ final class HBaseCredentialsManager private() extends Logging {
...
-    new TokenInfo(expireTime, conf, token)
+    val kerberosLoggedUGI = UserGroupInformation.loginUserFromKeytabAndReturnUGI(principal, keytab)
+    kerberosLoggedUGI.doAs(new PrivilegedExceptionAction[TokenInfo] {
+      override def run(): TokenInfo = {
+        val token = TokenUtil.obtainToken(conf)
+        val tokenIdentifier = token.decodeIdentifier()
+        val expireTime =
+          expectedExpireTime(tokenIdentifier.getIssueDate, tokenIdentifier.getExpirationDate)
+        logInfo(s"Obtain new token with expiration time $expireTime")
+        new TokenInfo(expireTime, conf, token)
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseCredentialsManager.scala]: @weiqingy Please make sure we change this to accommodate review comments from @bikassaha about not overloading expireTime; but rely on renewTime and expireTokenInfoTime.
Also about caching the serialized token and not token itself.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseCredentialsManager.scala [hunk 4]
```diff
@@ -150,5 +163,16 @@ final class HBaseCredentialsManager private() extends Logging {
...
+
+  def get(sparkConf: SparkConf): HBaseCredentialsManager = {
+    if (_hbaseCredentialManager == null) {
+      HBaseCredentialsManager.synchronized {
+        if (_hbaseCredentialManager == null) {
+          _hbaseCredentialManager = new HBaseCredentialsManager(sparkConf)
+        }
+      }
+    }
+    _hbaseCredentialManager
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseCredentialsManager.scala]: This change is not required - lazy val does the exact same thing without all the code.
In addition, HBaseCredentialsManager must be created/initialized only when configuration enables it.

**Comment by weiqingy** [comment]
@jerryshao I just reverted the PR, but we can always merge it in after #120 is in and the comments are resolved.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseCredentialsManager.scala (hunk 3 above)
Comment (weiqingy) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseCredentialsManager.scala]: Sure. I will update the PR to caching the serialized token instead.

**Comment by jerryshao** [comment]
Sure, thanks.

**Comment by weiqingy** [comment]
Hi, @jerryshao PR#120 is merged. Could you please commit the PR again? Thanks.

**Comment by jerryshao** [comment]
Will do it. Currently stuck on the some Spark issues, will do it after that.

**Pull request merged**